cue/cmd: Prevent exec.Run fatal errors for non-zero exit codes

nickfiggins opened the pull request:
This updates the logic in exec.Run to not return an error if a command returns a non-zero exit status. As a result, the success field can be used to determine if a command succeeded or failed.

Any errors that aren't related to the command failing to run or not completing successfully are still returned.

Fixes #2632

nickfiggins commented:
been learning cue and this issue sounded interesting, so wanted to give it a shot. Let me know if I misunderstood the intent here at all 😃

mvdan commented on pkg/tool/exec/exec.go:
```diff
@@ -80,15 +80,19 @@ func (c *execCmd) Run(ctx *task.Context) (res interface{}, err error) {
+		return update, nil
+	}
+
+	if exit := (*exec.ExitError)(nil); errors.As(err, &exit) {
```
I've never seen this single-line style for As before, interesting :) seems fine to me to be clear.

mvdan commented on cmd/cue/cmd/testdata/script/cmd_errcode.txtar:
```diff
@@ -16,4 +19,5 @@ command: errcode: {
 		kind:   "exec"
 		cmd:    "ls --badflags"
 		stderr: string // suppress error message
+		success: true
```
Interesting, so any users relying on the current behavior where `cue cmd foo` fails if an `exec.Run` command returns a non-zero exit code would need to add a line like this.

That is technically a breaking change, and I wonder how many users it would affect in practice. Then again, the `success` field can't be of any use without a fix like this, like Roger pointed out in the original bug report, so I think we want to make the change anyway.

One thought outloud is that we could later also consider something like `exec.MustRun` which would be a shorthand for `exec.Run & {success: true}`.

rogpeppe replied:
Another possibility might be to have another field that can be explicitly set to allow a failed command. It would default to the current behaviour.

For example, `allowFailure: *false | bool`, or maybe `mustSucceed: *true | bool`

Then if someone wants to allow a failed command they can do `allowFailure: true` or `mustSucceed: false` and the `success` field will indicate if the command did actually fail.

I think I might prefer this approach as not only does it preserve compatibility, but it also means that the default behaviour is to fail when a command fails, which seems like it might be the right default.

nickfiggins replied:
Ah yeah - that makes sense to me in order avoid the breaking change and also keep the same default.  Happy to add that in if there's alignment here 🙂 

One thing I am thinking about is assuming we would still want to have a fatal error when `cmd.Run` returns an error that isn't of type `os.ExitError` (as done in this PR currently), I'm thinking `mustSucceed: false` might be clearer than `allowFailure: true` since the latter may make it seem like that command would never fail?  

However, if we want to disregard all errors when a flag is set then maybe setting `allowFailures: true` would make more sense. Let me know what you all think / if there's a preference around that.

jpluscplusm replied:
Apologies if I'm missing something,  but may I suggest that the field not be a boolean, but an int instead - representing the command's exit code? 

That way, a default of "0" could be applied to keep the current backwards-compatible behaviour, whilst allowing the user control over which non-zero exit codes are permitted, via standard CUE number bounds, or a disjunction, or any other CUE-y mechanism.

mvdan replied:
@nickfiggins please hold off on updating this PR - we had a brief internal discussion on what kind of API change we should make here, so I'll polish up those notes and share early next week :)

myitcv replied:
> Apologies if I'm missing something, but may I suggest that the field not be a boolean, but an int instead - representing the command's exit code?

I believe the reason the `tool/exec` API does not expose the exit code is the same reason it doesn't appear in the https://pkg.go.dev/os/exec API: namely the behaviour around exit codes is OS-specific. @mvdan might well be able to offer more thoughts/colour on that point.

mvdan replied:
Correct; exit codes are relatively common in Unix-like and POSIX-compliant systems, and Go does seem to support them on all platforms including Windows if you reach for the lower-level https://pkg.go.dev/os#ProcessState.ExitCode API, but they [aren't a portable standard](https://stackoverflow.com/questions/1538884/what-standard-application-return-exit-codes-should-an-application-support) that is well understood and should replace `success: bool`.

That said, we could still expose the exit code alongside the success boolean, if that is useful to some users who for example run on Linux. If anyone has need for that, please raise an issue and we can discuss a change.

mvdan reviewed (commented):
Thanks for this patch! This looks great in terms of implementation and testing :)

Let me run it by the others to confirm whether we are OK with the user-facing API change.

mvdan commented:
@nickfiggins apologies for the delay - we have discussed a number of options and to what degree we want to keep backwards compatibility here. `Run` is documented as follows, which would indicate that a command run isn't required to succeed:

> success is set to true when the process terminates with a zero exit code or false otherwise. The user can explicitly specify the value force a fatal error if the desired success code is not reached.

However, the currently implemented behavior has been in place for years, and some of those users may depend on it, so asking them to switch to a new API or set an option would be unfortunate and might catch some users by surprise. Moreover, going ahead, we think that the default _should_ be to require a command to succeed to continue the execution of `cue cmd`, much like CI systems like GitHub Actions run the Bash shell with `set -o errexit`, where any command failure stops the entire script.

So, following @rogpeppe's thoughts above, we're thinking of adding a boolean flag:

```
// Run continues to result in a fatal error if success==false, and we add the "mustSucceed" field.
Run: {
  // rest of fields unchanged, including "success"

  // mustSucceed indicates whether a command must succeed, in which case success==false results in a fatal error.
  // This option is enabled by default, but may be disabled to control what is done when a command execution fails.
  mustSucceed: bool | *true
}
```

Existing users would continue to do exactly the same, for example:

```
cannotFail: exec.Run & {cmd: "foo"}
```

And then, users who want to use the "success" boolean as either true or false can set "mustSucceed" to false:

```
canFail: exec.Run & {cmd: "bar", mustSucceed: false}
if canFail.success {
  // logic based on the resulting status
}
```

Let me know if the above is clear. If you are able to update the PR this week, we will attempt to include this change in the upcoming v0.7.0 release. Thanks!

nickfiggins committed:
pkg/tool/exec: add a Run.mustSucceed parameter

Currently, exec.Run results in a fatal error if the command returns
a non-zero exit status, meaning that the "success" field can only ever
be used by other CUE expressions when it is set to true.

Add a mustSucceed boolean parameter, defaulting to true to keep the
default behavior fully backwards compatible.
When set to false, exec errors will allow "cue cmd" to continue
with the "success" field set to false.

Fixes #2632.

Signed-off-by: Nick Figgins <figginsn@gmail.com>
Co-authored-by: Daniel Martí <mvdan@mvdan.cc>
Change-Id: If52e16549dddfbc74db514cd562cf9fefc26f9f0

nickfiggins commented:
@mvdan Thanks for all the detail and explanation, just pushed the changes 🙂 . Let me know if there's any issues

mvdan reviewed (approved):
Thank you! I think you can slightly simplify two bits of the code, but since we want to release v0.7.0-rc.1 today with this change, I hope it's okay with you if I make those changes as I import the PR into Gerrit for merging.

mvdan commented:
Looking more closely at your latest commit:

> Any errors that aren't related to the command failing to run or not completing successfully are always still returned.

I tend to disagree there; it still sounds useful to use `mustSucceed: false` to capture other types of exec errors, such as when a program isn't available like `cmd: "non-existent-program"`. It would also be rather surprising if `mustSucceed: false` allowed some failures to continue evaluation, but others not.

I'll tweak your code so that it treats non-ExitError errors with the same logic, and appends their error strings to `Run.stderr`, so that they may be showed to the end user or used in some way by the CUE logic. We could later add more fields to `Run` to differentiate "command failed to start" versus "command started and then exited with a non-zero exit code" errors, but for now that doesn't seem particularly important.

I'll add myself as co-author in the commit, for the sake of clarifying that there were non-trivial tweaks. I'll also push another commit here, so that one can diff between the two. I'm again hoping that this is OK with you, for the sake of including the fix in v0.7 :)

mvdan commented:
Done; you can look through https://github.com/cue-lang/cue/compare/5dd19902850e7d41fcd241c9d9ba539d992627f5..a454651883a739e0109c3399c7197e2908c11c58 to see my changes. There is a bit of other noise unfortunately, since I also rebased on master.

Importing to Gerrit now, for one last round of review.

nickfiggins commented:
> I tend to disagree there; it still sounds useful to use `mustSucceed: false` to capture other types of exec errors, such as when a program isn't available like `cmd: "non-existent-program"`. It would also be rather surprising if `mustSucceed: false` allowed some failures to continue evaluation, but others not.
> 

That totally makes sense, I was going to ask for clarity around that but figured there wasn't much time if it were to be included in the next release 🙂 .

> Done; you can look through https://github.com/cue-lang/cue/compare/5dd19902850e7d41fcd241c9d9ba539d992627f5..a454651883a739e0109c3399c7197e2908c11c58 to see my changes. There is a bit of other noise unfortunately, since I also rebased on master.
> 
> Importing to Gerrit now, for one last round of review.

Ah I see, this makes sense. Thank you!

mvdan commented:
I ended running out of time to cut a release today, so it will have to be monday. Perhaps there wasn't a need for me to rush merging this after all :) In any case, if you have any further thoughts about the design, or want to send a follow-up PR with any other changes, by all means please do!
